.github/workflows/nightly.yml

```diff
@@ -0,0 +1,229 @@
+    # https://github.community/t/path-to-action-in-the-same-repository-as-workflow/16952/2
+    - name: Run ${{ matrix.test-kind }} tests ('${{ matrix.test-marker }}')
+      uses: ./.github/workflows/actions/run-tests
+      with:
+        test-kind: ${{ matrix.test-kind }}
+        test-marker: ${{ matrix.test-marker }}
+    # Currently GitHub workflow cannot call an action from another action
+    # https://github.community/t/call-an-action-from-another-action/17524
+    # Else this shared step can also be move to the local action: .github/workflows/actions/run-tests
+    - name: Upload Code Coverage
+      uses: actions/upload-artifact@v2
+      with:
+        name: code-cov
+        path: .coverage*
+
+###############################################
+################# SPARK-BUILD #################
+###############################################
+  build-spark:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        java: [8]
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit', 'smoke'] 
+        # pytest markers configured in tox.ini. See https://docs.pytest.org/en/6.2.x/example/markers.html
+        test-marker: ['spark and not gpu']
+        include:
+          # Integration tests needs a more powermachine with more memory. GitHub-hosted agents only have 7GB memory.
+          - os: self-hosted
+            python: 3.6
+            test-kind: 'integration'
+            test-marker: 'not spark and not gpu'
+
+        
+
+    steps:
+    - uses: actions/checkout@v2
+    ################# Install spark dependencies (java) #################
+    - name: Setup Java JDK
+      uses: actions/setup-java@v2.1.0
+      with:
+        java-version: ${{ matrix.java }}
+        distribution: 'adopt'
+    
+    ################# Run Python tests #################
+    - name: Use Python ${{ matrix.python }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python }}
+
+    - name: Run ${{ matrix.test-kind }} tests ('${{ matrix.test-marker }}')
+      uses: ./.github/workflows/actions/run-tests
+      with:
+        test-kind: ${{ matrix.test-kind }}
+        test-marker: ${{ matrix.test-marker }}
+      
+    - name: Upload Code Coverage
+      uses: actions/upload-artifact@v2
+      with:
+        name: code-cov
+        path: .coverage*
+
+###############################################
+################# GPU-BUILD #################
+###############################################
+  build-gpu:
+    runs-on: [self-hosted, Linux, gpu] # this is a union of labels to select specific self-hosted machine
+    needs: static-analysis
+    strategy:
+      matrix:
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit', 'smoke', 'integration'] 
```

Comment:
in the other builds we have only `test-kind: ['unit', 'smoke']`, what is the difference?

Reply:
This difference is due to how we need to make an exception for running integrations in self-hosted machine. 

Here in the GPU-build, there aren't problem running all of the 3 different test-kinds in the self-hosted GPU machine, so there's no `include` clause in the matrix. In the SPARK-build, some of the the integration & unit tests will fail running on the github-hosted machines (i.e. `os: [ubuntu-latest]`). So I added a `include` clause to make rooms for such exception.

Reply:
do we know when GitHub has plans to add GPU hosted machines?

Reply:
Unfortunately, I don't have much information on that. For folks I asked, they don't have visibility in these feature planning cycles. My assumption is that CI workload requiring GPU resources is quite rare and those who do can be refer to self-hosting their GPU resources. There might not be enough incentives in providing managed GPU resources. Also, recently hackers have been [exploiting these small GitHub Action SKUs to mine bitcoins](https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/), so security could be big concern if they offer GPU machine.